Issue 2283: Add type-review support for SPARQL query results from Graph.query

ajnelson-nist opened the pull request:
<!--
Thank you for your contribution to this project. This project has no formal
funding or full-time maintainers and relies entirely on independent
contributors to keep it alive and relevant.

This pull request template includes some guidelines intended to help
contributors, not to deter contributions. While we prefer that PRs follow our
guidelines, we will not reject PRs solely on the basis that they do not, though
we may take longer to process them as in most cases the remaining work will
have to be done by someone else.

If you have any questions regarding our guidelines, submit the PR as is
and ask.

More detailed guidelines for pull requests are provided in our [developers
guide](https://github.com/RDFLib/rdflib/blob/main/docs/developers.rst).

As a reminder, PRs that are smaller in size and scope will be reviewed and
merged quicker, so please consider if your PR could be split up into more than
one independent part before submitting it, no PR is too small. The maintainers
of this project may also split up larger PRs into smaller more manageable PRs
if they deem it necessary.

PRs should be reviewed and approved by at least two people other than the
author using GitHub's review system before being merged. Reviews are open to
anyone, so please consider reviewing other open pull requests as this will also
free up the capacity required for your PR to be reviewed.
-->

# Summary of changes

<!--
Briefly explain what changes the pull request is making and why. Ideally this
should cover all changes in the pull request as the changes will be reviewed
against this summary to ensure that the PR does not include unintended changes.

Please also explicitly state if the PR makes any changes that are not backwards
compatible.
-->

# Checklist

<!--
If an item on this list doesn't apply to your pull request, just remove it.

If, for some reason, you can't check some items on the checklist, or you are
unsure about them, submit your PR as is and ask for help.
-->

- [x] Checked that there aren't other open pull requests for
  the same change.
   - It's possible [PR 2301](https://github.com/RDFLib/rdflib/pull/2301) and/or [PR 2267](https://github.com/RDFLib/rdflib/pull/2267) will have some impact on this implementation.  That's a guess from seeing the wide-ranging set of type signature complaints encountered in this PR's initial patch.
- [x] Added tests for any changes that have a runtime impact.
   - [x] Added tests for `SELECT` and `CONSTRUCT` queries
   - [x] Added tests for `DESCRIBE` and `ASK` queries
- [ ] Checked that all tests and type checking passes.  *(Submitter: Known to not pass yet; will pass before PR is requested to be merged.)*
- For changes that have a potential impact on users of this project:
  - [ ] Updated relevant documentation to avoid inaccuracies.
  - [ ] Considered adding additional documentation.
  - [ ] Considered adding an example in `./examples` for new features.
  - [ ] Considered updating our changelog (`CHANGELOG.md`).
- [x] Considered granting [push permissions to the PR branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork),
  so maintainers can fix minor issues and keep your PR up to date.

ajnelson-nist committed:
Start type hinting mechanisms for Graph.query

This patch is known to be an incomplete implementation.  Running
`mypy test/test_typing.py` before this patch raised 0 type errors.
After just adding the two new tests, 3.  After starting the subclasses
of Query, 30.

References:
* https://github.com/RDFLib/rdflib/issues/2283

Signed-off-by: Alex Nelson <alexander.nelson@nist.gov>

ajnelson-nist committed:
Update disclaimer

Signed-off-by: Alex Nelson <alexander.nelson@nist.gov>

ajnelson-nist committed:
Merge branch 'main' into issue-2283

ajnelson-nist committed:
Merge branch 'main' into issue-2283

aucampia commented:
PRs to V6 is closed until further notice. See this for more details:
- https://github.com/RDFLib/rdflib/discussions/2395

aucampia commented:
> PRs to V6 is closed until further notice. See this for more details:
> 
> * [RDFLib V6 and V7 #2395](https://github.com/RDFLib/rdflib/discussions/2395)

We will be open for PRs again once this is resolved:

- https://github.com/RDFLib/rdflib/pull/2402

ajnelson-nist committed:
Merge branch 'main' into issue-2283

nicholascar committed:
Merge branch 'main' into issue-2283

nicholascar commented:
@ajnelson-nist do you want to continue with this PR not that v7.x is out and we are merging PRs again?

ajnelson-nist committed:
Merge branch 'main' into issue-2283

ajnelson-nist committed:
Fix sort order

Signed-off-by: Alex Nelson <alexander.nelson@nist.gov>

ajnelson-nist committed:
Switch to Query.prepare classmethod

Potentially, adding an assert() addresses the subclass type review,
per `@classmethod` docs:

> ... If a class method is called for a derived class, the derived class
> object is passed as the implied first argument.

This patch also reverts the first-draft change to `Graph.query`.

References:
* https://docs.python.org/3/library/functions.html#classmethod
* https://github.com/RDFLib/rdflib/issues/2283#issuecomment-1485923912

Suggested-by: Iwan Aucamp <aucampia@gmail.com>
Signed-off-by: Alex Nelson <alexander.nelson@nist.gov>

ajnelson-nist committed:
Format

ajnelson-nist committed:
Draft remaining tests

References:
* https://github.com/RDFLib/rdflib/issues/2283

Signed-off-by: Alex Nelson <alexander.nelson@nist.gov>

ajnelson-nist commented:
I've given this a nudge.  Following @aucampia 's suggestion looks promising, but it runs into a circular import issue bringing in `prepareQuery` at the point he'd suggested.

The class `rdflib.query.Result` caught my eye as a potential next point of inspection.

ajnelson-nist commented:
I'm pausing working on this for the time being.  Suggestions are welcome.
